Sync GitHub app entry metadata with default version

kathy-t opened the pull request:
**Description**
The metadata for the [workflow in the ticket description](https://staging.dockstore.org/workflows/github.com/broadinstitute/viral-pipelines/kraken2_build:master?tab=info) was not properly synced with the default version's (master) metadata. I looked further into it and found a [workflow version (v2.1.8.0)](https://staging.dockstore.org/workflows/github.com/broadinstitute/viral-pipelines/kraken2_build:v2.1.8.0?tab=info) that did match the entry level description. 

My guess for what happened is:
1) The GitHub app workflow was registered.
2) The user picked master as the default version in the UI, which synced the workflow-level metadata with this default version. The master version's description was the README at that point in time because the [primary descriptor didn't have a description](https://github.com/broadinstitute/viral-pipelines/blob/cda126d0bc653ac0ff062b9816f2375f06ba9c15/pipes/WDL/workflows/kraken2_build.wdl).
3) The master branch was modified so that the [primary descriptor has a description](https://github.com/broadinstitute/viral-pipelines/blob/master/pipes/WDL/workflows/kraken2_build.wdl#L7). The master version's description is now different from the workflow's description. The GitHub release did not sync the workflow metadata with the updated default version metadata.

I also found that although setting `latestTagAsDefault: true` sets the latest tag as the default version, it didn't sync the workflow metadata with the default version either.

This PR syncs the GitHub app entry metadata with the default version's metadata every time there's a GitHub release. This syncing problem only occurs with GitHub app entries from my testing.

**Issue**
#4729 
https://ucsc-cgl.atlassian.net/browse/DOCK-2065

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future

kathy-t commented:
Failed CircleCI tests are marked as flaky, so will re-run them later

coverbeck reviewed (commented):
> The user picked the master branch as the default version in the UI, but the workflow metadata was not synced with this new default version so the workflow metadata is out of sync with the new default version.

Wouldn't it be that:

1. The user picked master as the default in the UI, which does look like it [syncs the entry metadata](https://github.com/dockstore/dockstore/blob/1b9b860f9aab7f57fde5a98863e06391528e2f16/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L536)
2. The user then updated the primary descriptor with a description, which got pushed via GitHub apps, but didn't sync the entry metadata?

Or on second reading, maybe that's what you meant, but I first read your comment as saying changing the default version in the UI doesn't update the metadata.

Also, is it possible this bug occurs with legacy workflows? I skimmed and it looks like there is code for when you [refresh a workflow](https://github.com/dockstore/dockstore/blob/1b9b860f9aab7f57fde5a98863e06391528e2f16/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L396), but didn't find it if you refresh a version (didn't look too hard though).

kathy-t commented:
> > The user picked the master branch as the default version in the UI, but the workflow metadata was not synced with this new default version so the workflow metadata is out of sync with the new default version.
> 
> Wouldn't it be that:
> 
>     1. The user picked master as the default in the UI, which does look like it [syncs the entry metadata](https://github.com/dockstore/dockstore/blob/1b9b860f9aab7f57fde5a98863e06391528e2f16/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L536)
> 
>     2. The user then updated the primary descriptor with a description, which got pushed via GitHub apps, but didn't sync the entry metadata?
> 
> 
> Or on second reading, maybe that's what you meant, but I first read your comment as saying changing the default version in the UI doesn't update the metadata.

Oh yes, that's what I meant :facepalm: and that's what the test I added reflects. Changing the default version in the UI updates the metadata as expected. Updated my PR description

> Also, is it possible this bug occurs with legacy workflows? I skimmed and it looks like there is code for when you [refresh a workflow](https://github.com/dockstore/dockstore/blob/1b9b860f9aab7f57fde5a98863e06391528e2f16/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L396), but didn't find it if you refresh a version (didn't look too hard though).

The function that you linked takes an [optional version ID](https://github.com/dockstore/dockstore/blob/1b9b860f9aab7f57fde5a98863e06391528e2f16/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L346) which is provided when [refreshing a version](https://github.com/dockstore/dockstore/blob/1b9b860f9aab7f57fde5a98863e06391528e2f16/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L332) so this bug shouldn't affect refreshing a version

kathy-t commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:
```diff
@@ -530,6 +530,7 @@ private boolean createWorkflowsAndVersionsFromDockstoreYml(List<? extends Workfl
 
                         Workflow workflow = createOrGetWorkflow(workflowType, repository, user, workflowName, subclass, gitHubSourceCodeRepo);
                         WorkflowVersion version = addDockstoreYmlVersionToWorkflow(repository, gitReference, dockstoreYml, gitHubSourceCodeRepo, workflow, defaultVersion, yamlAuthors);
+                        workflow.syncMetadataWithDefault();
```
> It seems like we should only be syncing if the default version changed or was updated. I suppose it's harmless to do it otherwise, and it will update ones we missed, but still...

I didn't check if the default version changed or was updated because it didn't look like we did any checking for other places where this function is called

coverbeck replied:
> I didn't check if the default version changed or was updated because it didn't look like we did any checking for other places where this function is called

I feel like this use case is a little different. At a glance, it looks like the other cases it does the sync are when:
1. It's refreshing the entire workflow/tool
2.  Explicitly setting the default version
3. Saving a hosted entry's version; but saving it, makes it the default, so this is a variant of 2.

In this code, we were notified by GitHub of an update to a specific branch/tag, and we should know if the corresponding Dockstore version being updated is the default or not.

But it won't hurt to do this and the object is loaded into memory, so it shouldn't have a performance impact.

denis-yuen reviewed (approved):
thanks for the substantive test!
